metrics.go

```diff
@@ -1,23 +1,51 @@
 package f3
 
 import (
+	"container/list"
 	"context"
+	"sync"
 	"time"
 
+	"github.com/filecoin-project/go-f3/gpbft"
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	"go.opentelemetry.io/otel"
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/metric"
 )
 
 var meter = otel.Meter("f3")
+var sampleSets = struct {
+	messages       *sampleSet
+	justifications *sampleSet
+}{
+	// The the max size of 25K is based on network size of ~3K and 5 phases of GPBFT,
+	// where the vast majority of instances should complete in one round. That makes
+	// up ~15K messages per instance. We may observe messages from previous or future
+	// rounds, hence an additional capacity of 10K, making a total of 25K messages.
+	//
+	// Although, the number of observable messages with justification are expected to
+	// be less than the total number of observed messages we use the same 25K bounded
+	// size for justifications. The memory footprint of the additionally stored
+	// samples is negligible in return for a larger sample size and ultimately a more
+	// accurate measurement considering justification signature validation is more
+	// expensive than message signature validation.
+	//
+	// Since sampleSet caps the max size of cache keys to 96 bytes (the length
+	// of BLS signatures) the total memory footprint for 50K samples should be be
+	// below 10MB (5MB * 2 due to bookkeeping overhead).
+	messages:       newSampleSet(25_000),
+	justifications: newSampleSet(25_000),
```

Comment:
1. I'd consider doubling this, just for better metrics.
2. It shouldn't be that hard to benchmark this. Fill up a set then call `runtime.ReadMemStats`.

Reply:
We could bump this later if numbers come back high, unless you think 25K is too small to begin with?

Reply:
Nah, should be fine.

Reply:
Thanks to flippity flopping i think we effectively get uniqueness check over 2X the max size with the same (if not slightly less) memory footprint 🎉